rfc-141-application-healthchecks.md

```diff
@@ -0,0 +1,224 @@
+performed some time in the next two months.  They all share the
+property that if one instance reports a failure, all will.  This makes
+them unsuitable for load balancing purposes.
+
+Even if we don't make our healthchecks suitable for load balancing, as
+GOV.UK moves towards continuous deployments, we will end up in
+situations where an automatic deployment is aborted because of
+something unrelated to the change being deployed.  That's not good.
+
+A healthcheck should check that the instance is running and isn't
+completely broken.  That's all.
+
+## Proposal
+
+My proposal comes down to three parts:
+
+1. Remove the "warning" state of `GovukHealthcheck`, so a healthcheck
+   result is either "ok" or "critical".
+2. Make `GovukHealthcheck` serve an HTTP status of 500 for a
+   "critical" result.
+3. Add a separate healthcheck for liveness, to use with the ECS
+   container lifecycle.
+
+But we can't make this change right now, we have a migration path to
+follow first.
+
+### Remove the "warning" state
+
+There are four possible semantics we could assign the "warning" state:
+
+| # | State    | Allows automatic deployments? | Allows requests to be sent to the instance? |
+| - | -------- | ----------------------------- | ------------------------------------------- |
+|   | ok       | yes | yes |
+| 1 | warning  | yes | yes |
+| 2 | warning  | yes | no  |
+| 3 | warning  | no  | yes |
+| 4 | warning  | no  | no  |
+|   | critical | no  | no  |
+
+- If we go for option 1, then "warning" is the same as "ok".
+- If we go for option 2, then "warning" will let us deploy unusable releases.
+- If we go for option 3, then "warning" will block deployments.
+- If we go for option 4, then "warning" is the same as "critical".
+
+The most sensible option is 3, which is our current behaviour.
+
+But does it really gain us anything over having separate alerts for
+the specific condition we need to know about?
+
+By removing the "warning" state, we will remove some spurious alerts
+which don't add value, and add more specific alerts for ones which do.
+
+### Make `GovukHealthcheck` return a 500 status for critical failure
+
+This is a requirement for `GovukHealthcheck` to be usable with load
+balancers.
+
+### Have separate healthchecks for liveness and readiness
+
+If `GovukHealthcheck` checks whether an instance is ready to handle
+requests, and takes it out of the load balancer target group if not,
+then it is a *readiness* check.
+
+There is also value in having a separate *liveness* check, which
+simply checks if the instance is running and reachable, and doesn't
+check any of its dependencies:
+
+```ruby
+get "/healthcheck", to: proc { [200, {}, []] }
+get "/healthcheck/ready", to: GovukHealthcheck.rack_response(...)
+```
+
+When we have replatformed to ECS, this healthcheck can be used to
+determine if a container needs to be automatically restarted, for
+example if the process has crashed.
```

Comment:
Thanks for considering ECS! The `/healthcheck/ready` endpoint would be _perfect_ for [ECS container health checks](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_HealthCheck.html).

Am I correct in thinking that the endpoint `/healthcheck/ready` indicates liveness _and_ readiness, and that the `/healthcheck` endpoint indicates that an app can reach its dependencies?

Do these definitions make sense?

**Healthy**: The application can reach its critical dependencies (e.g. frontend can reach content-store)
**Ready**: The application has finished booting up (e.g. router has loaded routes into memory)

Some examples of health and liveness checks would be useful, to make the distinction between what should go into each health check clearer. It would be great to define how each health check should be used, too, to give us an idea about how to use them in ECS.

I am not 100% clear on the value of the "Healthy"/dependency checks (for use by a load balancer or alerting).

I would recommend removing the Healthy checks, and just keep the the readiness checks for load balancing. Because:

* Dependency health checks will mostly fail due to problems outside of the load balancer's control, and won't be a useful signal during deployments
* We don't plan on using health checks for alerting in ECS. Our intent in ECS is to [change how we do alerting](https://docs.google.com/document/d/1unWtbmh2g2HtXbCsFQ0GSbG1LIVraGsq6DQwX24mTVg/edit#heading=h.7qovbilab9jq) and avoid alerting on health checks. I would prefer that we discover a dependency is failing by letting it fail and alerting on it at a higher level (e.g. 5xx rate for a service), rather than via a health check.

Forgive me if this was already your intent here!

You wrote that good health checks indicate "that the instance is running and isn't completely broken". I reckon you've covered that well with the readiness health check. Whereas you wrote that bad health checks indicate "problems outside of the instance", which is probably pretty much what the "Healthy"/dependency checks are indicating.

I found dependency healthchecks a bit frustrating when trying to bring up apps in ECS, since containers would be stopped because a dependency wasn't running. Health checking dependencies (at any level of load balancing) prevented apps from degrading service gracefully (i.e. 50% of pages will work if half of the dependencies go down).

In the past I think we've actually been rescued during incidents from our health checks that check dependencies by AWS ALB's default behaviour. If all targets have failing health checks, then the health checks are ignored by the ALB - handy when a DB is down and you don't want to take down all the apps that use it, too!

> If all targets fail health checks at the same time in all enabled Availability Zones, the load balancer fails open. The effect of the fail open is to allow traffic to all targets in all enabled Availability Zones, regardless of their health status. [source](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html)

Reply:
> Am I correct in thinking that the endpoint `/healthcheck/ready` indicates liveness and readiness, and that the `/healthcheck` endpoint indicates that an app can reach its dependencies?

I was actually thinking:

- `/healthcheck` *just* checks that the process is running and reachable over HTTP.  It's just a static route returning a 200, with no additional logic.
- `/healthcheck/ready` checks that the app is ready to serve requests (what that means depends on the app: eg, able to reach backing services, or has loaded routes into memories).

The `/healthcheck` is useful for container-level checks, because all the container runtime cares about is that the app runs.  So this is what ECS would use.

The `/healthcheck/ready` is useful for load balancer-level checks, because the load balancer cares, in addition, that the app is able to serve requests.  So this is what ALBs would use.

> I would prefer that we discover a dependency is failing by letting it fail and alerting on it at a higher level (e.g. 5xx rate for a service), rather than via a health check.

I like alerting on user-visible things, but what about the continuous deployment smoke testing use-case?  If a new version of an app has a bug which prevents it talking to its database, we'd want to catch that before it starts serving actual traffic - which a readiness healthcheck ensuring it can successfully do so gives us.

Reply:
Thanks for clarifying! This sounds reasonable. I think this is fine provided the readiness health checks fail open, and the ALB checks are less aggressive about removing instances than the container checks, to avoid dependencies becoming 'hard' dependencies. In ECS we'll be using Envoy rather than ALBs for health checks, luckily they also fail open ([panic mode!](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/panic_threshold)) if all health checks fail.